core/src/main/java/com/linecorp/armeria/internal/common/TimeoutScheduler.java

```diff
@@ -18,73 +18,192 @@
-    private Consumer<TimeoutController> pendingTimeoutTask;
+    private volatile TimeoutFuture whenTimingOut;
+    // Updated via whenTimedOutUpdater
     @Nullable
-    private EventLoop eventLoop;
+    private volatile TimeoutFuture whenTimedOut;
+    // Updated via pendingTimeoutTaskUpdater
     @Nullable
-    private TimeoutController timeoutController;
+    private volatile Runnable pendingTimeoutTask;
+    // Updated via pendingTimeoutNanosUpdater
+    @SuppressWarnings("FieldMayBeFinal")
+    private volatile long pendingTimeoutNanos;
+
+    private volatile boolean initialized;
 
     public TimeoutScheduler(long timeoutNanos) {
         this.timeoutNanos = timeoutNanos;
+        pendingTimeoutNanos = timeoutNanos;
+    }
+
+    /**
+     * Initializes this {@link TimeoutScheduler}.
+     */
+    public void init(EventExecutor eventLoop, TimeoutTask timeoutTask, long initialTimeoutNanos) {
+        if (!eventLoop.inEventLoop()) {
+            eventLoop.execute(() -> init(eventLoop, timeoutTask, initialTimeoutNanos));
+            return;
+        }
+
+        this.eventLoop = eventLoop;
+        this.timeoutTask = timeoutTask;
+        if (initialTimeoutNanos > 0) {
+            timeoutNanos = initialTimeoutNanos;
+        }
+
+        ensureInitialized();
+        if (timeoutNanos != 0) {
+            state = State.SCHEDULED;
+            timeoutFuture = eventLoop.schedule(this::invokeTimeoutTask, timeoutNanos, TimeUnit.NANOSECONDS);
+        }
+
+        Runnable pendingTimeoutTask;
+        for (;;) {
+            pendingTimeoutTask = this.pendingTimeoutTask;
+            if (pendingTimeoutTaskUpdater.compareAndSet(this, pendingTimeoutTask, null)) {
+                break;
+            }
+        }
+
+        if (pendingTimeoutTask != null) {
+            pendingTimeoutTask.run();
+        }
+        initialized = true;
+        final Runnable newlyAdded = this.pendingTimeoutTask;
+        if (newlyAdded != null) {
+            newlyAdded.run();
+        }
     }
 
     public void clearTimeout() {
+       clearTimeout(true);
+    }
+
+    public void clearTimeout(boolean resetTimeout) {
         if (timeoutNanos == 0) {
             return;
         }
 
-        final TimeoutController timeoutController = this.timeoutController;
-        timeoutNanos = 0;
-        if (timeoutController != null) {
+        if (initialized) {
             if (eventLoop.inEventLoop()) {
-                timeoutController.cancelTimeout();
+                unsafeClearTimeout(resetTimeout);
             } else {
-                eventLoop.execute(timeoutController::cancelTimeout);
+                eventLoop.execute(() -> unsafeClearTimeout(resetTimeout));
             }
         } else {
-            addPendingTimeoutTask(TimeoutController::cancelTimeout);
+            setPendingTimeoutNanos(0);
+            addPendingTimeoutTask(() -> unsafeClearTimeout(resetTimeout));
+        }
+    }
+
+    private boolean unsafeClearTimeout(boolean resetTimeout) {
```

Comment:
nit: `clearTimeout0`? we normally use unsafe for other purposes.

Reply:
Armeria commonly uses `unsafe` for Pool Objects. But I think we can use the name - `unsafe`, if a method should be used with caution.
Initially, I use `xxxTimeout0`. But readability was not good. Because lots of timeout methods are mixed in this class. So I wanted to easily distinguish with other safe methods.

Reply:
Then, what `unsafe` means here? I hardly figure out what the word is for. 🤔

Reply:
I intended that the method is not thread safe. Should be called in an event loop. If `unsafe` make you confused, let me remove the word from the method name. 😉